### DISCUSSION TITLE
log4j (v1) Custom logger appender in Mirth

### DISCUSSION THREAD

**Discussion by tom08zehn** [nextgenhealthcare/connect#5137] [Q&A] [answered]
I would like to create my own custom logger appender in Mirth.

Goal/idea: I want to handle every new line in the logger with a kind of custom callback function.

The following link shows how to do that in native Java: https://stackoverflow.com/questions/6072389/how-to-create-my-own-appender-in-log4j 

But in Mirth I guess I somehow need to make use of `JavaAdapter`. Could someone help me and create a simple example as a starting point?

Many thanks!

**Comment by jonbartels** [comment]
The odo appender might be a good example since its MC specific - https://forums.mirthproject.io/forum/mirth-connect/support/18801-new-feature-mirth-logging-replacement and https://github.com/odoodo/Mirth-MetaAppender

You can see that odo's MetaAppender extends a log4j FileAppender https://github.com/odoodo/Mirth-MetaAppender/blob/master/lu/hrs/mirth/MetaAppender.java

If you can share, I am curious what your use case is and what you want your custom logger to do.

**Reply by tom08zehn** [reply]
I have seen this MetaAppender "beast" before but it's not exactly what I need.

What I want: see the logs of a *single message* in dashboard's "Mapping" tab, see screenshot [here](https://github.com/nextgenhealthcare/connect/discussions/5135). I want to debug a message quickly in the GUI without the need to check logfiles on the command line or in a separate editor. I'm also planning to use this concept for automated testing.

My current appender implementation is using a StringWriter and seems to cause memory issues. Not sure if the root cause is the StringWriter or the appender itself.

I want to rewrite some parts of my code to prevent running out of memory.

My idea is to replace StringWriter with a regular string (stored in channelMap). But standard logger appenders need a Writer object... A custom appender could help here. Maybe one could also use StringBuilder instead of StringWriter and the custom appender would simply call the StringBuilder.append() method.

By the way @jonbartels: the MetaAppender is written in native Java. I'm aiming for a much simpler solution written in Mirth/Rhino JS. **That's why I asked if anybody could help me "translating" the Java examples into JS code.** I just need a starting point.

**Reply by tonygermano** [reply]
Your screenshot just links back to this discussion. This sounds messy. Can you just create your own log function and call that instead?

**Reply by tom08zehn** [reply]
@tonygermano Thanks, I corrected the link to the screenshot: [#5135](https://github.com/nextgenhealthcare/connect/discussions/5135)

No, I cannot use a custom logger as the "logger copy" should include *all* logs incl. those of code template library functions and custom jar libraries that I cannot rewrite.

The easiest to me seems to create a custom logger *appender* that simply adds every logged line to a channelMap variable of type string.

That's why I'd like to understand how to create it using Mirth/Rhino JS as shown [here](https://stackoverflow.com/questions/6072389/how-to-create-my-own-appender-in-log4j): **How do I extend `AppenderSkeleton` and override its `append()` method using `JavaAdapter`?**

In Java it would simply look as follows:

```js
public /*static*/ class MyAppender extends AppenderSkeleton {
    @Override
    protected void append(LoggingEvent event) {
        channelMap("__LOG__", event.getRenderedMessage());
    }

    public void close() {
    }

    public boolean requiresLayout() {
        return false;
    }

}
```

**Comment by tonygermano** [comment, marked as answer]
I can show you how to use JavaAdapter, but I still don't see how this is supposed to work. How would an appender know to capture log messages from custom jar libraries that were only generated from the current message? And if you're only worried about the current script and code templates, then you shouldn't need to use an appender as you can just hijack the logger variable and replace it with a custom javascript object (which can, in turn, call the original logger object along with any custom behavior you want.)

I'm also not a log4j expert, but I imagine that once you add an appender that there are still lingering references to it that are causing your memory issues where you might be required to also unregister it at the end of the script (probably in a giant try/finally in case an exception is thrown at some point.) Be aware that upgrading to log4j 2.x is on the roadmap, and it probably will not be compatible with this solution when that happens.

I did not test, but it should look something like this:
```js
var myAppender = new JavaAdapter(org.apache.log4j.AppenderSkeleton, {
  append: function append(event) {
    channelMap.put("__LOG__", event.getRenderedMessage())
  },
  requiresLayout: function requiresLayout() {
    return false
  }
})
```

I'm sure you are aware that this will only show the last log message since it is just replacing the previous one rather than accumulating them.

**Reply by tom08zehn** [reply]
> ... as you can just hijack the logger variable and replace it with a custom javascript object (which can, in turn, call the original logger object along with any custom behavior you want.)

Ah, now I understood what you mean. I guess I would need to override the methods `debug()`, `info()`, `warn()`, `error()` and maybe also `trace()` in my custom logger? **Any example how you would realize this?** I imagine something like a hook: any method/property is called for the original logger object and _only if_ name of method is `"info"` etc. then do something before. Is this possible?

Another goal is to find out if the appender or the unclosed `StringWriter` is causing the memory issue. That's why I still want to try the custom appender.

> I'm sure you are aware that this will only show the last log message since it is just replacing the previous one rather than accumulating them.

Yes sure, was just a simplified example.

Many thanks!

**Reply by tonygermano** [reply]
This was also not tested in mirth, so let me know if it needs any corrections. Should be easy to add any other log levels you need. If using an older version of mirth without `Array.from(arguments)` you should be able to replace it with `Array.prototype.slice.call(arguments, 0)`.

```js
// fn is a function that takes a log level as a string and an array of arguments used for logging
function augmentLogger(fn) {
    var originalLogger = logger
    
    function doLogging(level, args) {
        originalLogger[level].apply(originalLogger, args)
        fn(level, args)
    }
    
    return {
        originalLogger: originalLogger,
        info: function() {doLogging('info', Array.from(arguments))},
        error: function() {doLogging('error', Array.from(arguments))},
        debug: function() {doLogging('debug', Array.from(arguments))}
    }
}

var logs = []
$co('__LOGS__', logs)

logger = augmentLogger((level, args) => {
    logs.push([level, args[0]])
})


logger.info('message 1')
logger.debug('another message')
```
